backend/src/main/java/fk/prof/backend/model/assignment/WorkAssignmentSchedule.java

```diff
@@ -0,0 +1,211 @@
+    if(this.maxAcceptableDelayForWorkAssignmentInSecs < (this.minAcceptableDelayForWorkAssignmentInSecs * 2)) {
+      throw new IllegalArgumentException(String.format("Max acceptable delay for work assignment = %d" +
+          "should be at least be twice of min acceptable delay for work assignment = %d", maxAcceptableDelayForWorkAssignmentInSecs, minAcceptableDelayForWorkAssignmentInSecs));
+    }
+
+    this.referenceTimeInNanos = System.nanoTime();
+    long initialOffsetInNanos = windowStartToleranceInSecs * NANOS_IN_SEC;
+    int minOverlap = targetScheduleEntries / maxScheduleEntriesWithNoOverlap;
+
+    int currentWorkAssignmentIndex = 0;
+    for (int i = 0; (minOverlap > 0) && (i < maxScheduleEntriesWithNoOverlap); i++) {
+      long offsetInNanos = initialOffsetInNanos + (effectiveProfileDurationInSecs * i * NANOS_IN_SEC);
+      for(int j = 0; j < minOverlap; j++) {
+        this.entries.add(new ScheduleEntry(workAssignmentBuilders[currentWorkAssignmentIndex++], offsetInNanos));
+      }
+    }
+
+    int remainingAssignments = targetScheduleEntries % maxScheduleEntriesWithNoOverlap;
+    assert targetScheduleEntries == (currentWorkAssignmentIndex + remainingAssignments);
+    for (int i = 0; i < remainingAssignments; i++) {
+      long offsetInNanos = initialOffsetInNanos + (effectiveProfileDurationInSecs * i * NANOS_IN_SEC);
+      this.entries.add(new ScheduleEntry(workAssignmentBuilders[currentWorkAssignmentIndex++], offsetInNanos));
+    }
+
+    maxConcurrentlyScheduledEntries = minOverlap + (remainingAssignments > 0 ? 1 : 0);
+  }
+
+  /**
+   * Returns the maximum number of concurrently scheduled entries in this schedule
+   * @return
+   */
+  public int getMaxConcurrentlyScheduledEntries() {
+    return maxConcurrentlyScheduledEntries;
+  }
+
+  /**
+   * Supplies with a work assignment
+   * Returns null if:
+   * > no work assignments are pending
+   * > no work assignment ready to be handed out
+   * > recorder already assigned work (tied to aggregation window)
+   * > timeout while acquiring lock over queue
+   * > interrupted while waiting to acquire lock over queue
+   * > exception occurred while processing queue entries
+   * @return WorkAssignment or null
+   */
+  public Recorder.WorkAssignment getNextWorkAssignment(RecorderIdentifier recorderIdentifier) {
+    try {
+      boolean acquired = entriesLock.tryLock(100, TimeUnit.MILLISECONDS);
+      if(acquired) {
+        try {
+          if(this.assignedRecorders.contains(recorderIdentifier)) {
+            return null;
+          }
+
+          ScheduleEntry scheduleEntry;
+          while((scheduleEntry = this.entries.peek()) != null) {
+            ScheduleEntry.ScheduleEntryValue value = scheduleEntry.getValue((System.nanoTime() - referenceTimeInNanos), minAcceptableDelayForWorkAssignmentInSecs, maxAcceptableDelayForWorkAssignmentInSecs);
+            if(value.tooEarly) {
```

Comment:
I understand the general theme that metric integration should be done sooner than later but not sure if above is a valid metric for us. poll being too early is a valid usecase and will happen all the time. We will start assigning work as late as possible in the aggregation window.